Meeting presentation mode

oliverguenther opened the pull request:
https://community.openproject.org/projects/meetings-stream/work_packages/63634/activity?query_id=5640

oliverguenther committed:
Presentation mode for meetings

oliverguenther committed:
Use separate controller

oliverguenther committed:
Pass presentation_mode around so we can keep its state

oliverguenther committed:
Add feature flag

oliverguenther committed:
Add spec

oliverguenther committed:
Auto-refresh

oliverguenther committed:
Touch agenda item on outcome changes

This allows us to reload only on agenda item changes

oliverguenther committed:
Show elapsed time

oliverguenther committed:
Use current_id and action to find next item, rather than passing directly

This fixes the cases where the next item is resorted

mrmir reviewed (changes requested):
Great job! I really like it 👏🏽 

I noticed some things:

1. Presenting an empty meeting needs to either not be possible or handled more gracefully. It returns a NoMethodError right now
2. While editing an agenda item in presentation mode, if it is updated somewhere else, the current edit state is lost and the component is reloaded with the externally updated content. The reload shouldn't be happening at all when an item is being edited according to the AC
3. The UI doesn't exactly match the specs (presentation mode button is different, timer is in a different location, exit button, etc.). Don't know how important these are to fix for v1, but I want to point them out nonetheless
4. The counter of agenda items/slides at the bottom is not updated until you switch to the next slide. Should this also be updated automatically instead?
5. Keyboard nav works great, but should there also be a way to exit via Esc?
6. For longer text, the content is cut rather abruptly at the bottom. I don't know if we want to add a shadow effect or something? I quite like it as is, but maybe the designers think otherwise
7. There's a bit of a flicker if you navigate to the next item if it was added by someone else once you're already presenting the current item. It first goes to the old next item before reloading to the new one. Don't know if this needs to be touched for now

Everything except the first two points can be dealt with later too, I think

oliverguenther committed:
Handle empty meetings

oliverguenther committed:
Only update the show component, ensuring not replacing an active edit mode

oliverguenther commented:
> Presenting an empty meeting needs to either not be possible or handled more gracefully. It returns a NoMethodError right now

Fixed

> While editing an agenda item in presentation mode, if it is updated somewhere else, the current edit state is lost and the component is reloaded with the externally updated content. The reload shouldn't be happening at all when an item is being edited according to the AC

Fixed by having the auto-refresh only reload the show component part, this will prevent changing an active edit.  You still get an error flash that the content has been changed (same as in regular meeting)

> The UI doesn't exactly match the specs (presentation mode button is different, timer is in a different location, exit button, etc.). Don't know how important these are to fix for v1, but I want to point them out nonetheless

As discussed, waiting for Parimal to provide updates

> The counter of agenda items/slides at the bottom is not updated until you switch to the next slide. Should this also be updated automatically instead?

As discussed, this was okay to be out of sync until you navigate

> Keyboard nav works great, but should there also be a way to exit via Esc?

As discussed, this probably breaks with fullscreen mode anyway. I'll add that in a separate ticket to evalute

> For longer text, the content is cut rather abruptly at the bottom. I don't know if we want to add a shadow effect or something? I quite like it as is, but maybe the designers think otherwise

As discussed, this is as expected

> There's a bit of a flicker if you navigate to the next item if it was added by someone else once you're already presenting the current item. It first goes to the old next item before reloading to the new one. Don't know if this needs to be touched for now

This is coming from the turbo-prefetch on the links. We said we wanted to keep this for now, as it optimized for speed rather than correct content on quicker load.

mrmir commented:
@oliverguenther it's possible to present a template. I don't think that makes sense to keep

oliverguenther committed:
Avoid showing button for templates

oliverguenther commented:
> @oliverguenther it's possible to present a template. I don't think that makes sense to keep

Indeed, fixed and specced 👍

oliverguenther commented:
@lindenthal I fixed the PullPreview deployment, it's now available here: https://pr-20865-meetings-present-ip-3-75-191-4.my.preview.run/projects/demo-project/meetings/2

oliverguenther committed:
Avoid double respond_to

oliverguenther committed:
Add fullscreen

oliverguenther committed:
Add next/previous titles

oliverguenther committed:
Extract into multiple components

oliverguenther committed:
Restrict width to 900px + padding

oliverguenther committed:
Automatically update header and footer

oliverguenther committed:
Add spec for refresh

oliverguenther committed:
Update align flex-start

oliverguenther committed:
Update header spacing

oliverguenther committed:
Update state when presenting

Use separate start route so we can POST to it, indicating that it changes the meeting

mrmir reviewed (approved):
It looks good overall, but I have a few questions:

1. An empty section field is present when there are no sections in the meeting. Can this be hidden?
2. For long agenda items that need to scroll, would it be possible to keep the header fixed like the footer and only scroll the content in between?

oliverguenther committed:
Hide empty section

oliverguenther committed:
Sticky header

oliverguenther commented:
> It looks good overall, but I have a few questions:
> 
> 1. An empty section field is present when there are no sections in the meeting. Can this be hidden?
> 2. For long agenda items that need to scroll, would it be possible to keep the header fixed like the footer and only scroll the content in between?

Good point, fixed both and added to specification
